File: src/sagsins-node/src/main/java/com/sagin/model/Packet.java
package com.sagin.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;

@Data 
@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL) 
public class Packet {

    private String packetId;
    private String sourceUserId;
    private String destinationUserId;
    private String stationSource;
    private String stationDest;
    private String type;
    private String acknowledgedPacketId;
    private long timeSentFromSourceMs;
    private String payloadDataBase64;
    private int payloadSizeByte;
    private ServiceQoS serviceQoS;
    private int TTL;
    private String currentHoldingNodeId;
    private String nextHopNodeId;
    private List<String> pathHistory = new ArrayList<>();
    private List<HopRecord> hopRecords = new ArrayList<>();
    private double accumulatedDelayMs = 0.0;
    private int priorityLevel = 1;
    private boolean isUseRL = false;
    private double maxAcceptableLatencyMs = 150.0;
    private double maxAcceptableLossRate = 0.01;
    private boolean dropped = false;
    private String dropReason;
    private AnalysisData analysisData;

    /**
     * Giải mã payload từ Base64 sang chuỗi UTF-8.
     * @return Chuỗi đã giải mã hoặc thông báo lỗi.
     */
    @JsonIgnore 
    public String getDecodedPayload() {
        try {
            byte[] decodedBytes = Base64.getDecoder().decode(this.payloadDataBase64);
            return new String(decodedBytes, StandardCharsets.UTF_8);
        } catch (Exception e) {
            return "[Error decoding Base64]";
        }
    }
}

File: src/rl-router/model/Packet.py
from typing import List, Optional
from dataclasses import dataclass, field
from enum import Enum

class RoutingAlgorithm(str, Enum):
    DIJKSTRA = "Dijkstra"
    REINFORCEMENT_LEARNING = "ReinforcementLearning"

@dataclass
class Position:
    latitude: float
    longitude: float
    altitude: float

@dataclass
class BufferState:
    queue_size: int
    bandwidth_utilization: float  # 0.0 - 1.0

@dataclass
class RoutingDecisionInfo:
    algorithm: RoutingAlgorithm
    metric: Optional[str] = None
    reward: Optional[float] = None

@dataclass
class HopRecord:
    from_node_id: str
    to_node_id: str
    latency_ms: float
    timestamp_ms: float
    distance_km: float
    from_node_position: Optional[Position] = None
    to_node_position: Optional[Position] = None
    from_node_buffer_state: Optional[BufferState] = None
    routing_decision_info: Optional[RoutingDecisionInfo] = None
    scenario_type: Optional[str] = None
    node_load_percent: Optional[float] = None
    drop_reason_details: Optional[str] = None

@dataclass
class QoS:
    service_type: str
    default_priority: int
    max_latency_ms: float
    max_jitter_ms: float
    min_bandwidth_mbps: float
    max_loss_rate: float

@dataclass
class AnalysisData:
    avg_latency: float
    avg_distance_km: float
    route_success_rate: float
    total_distance_km: float
    total_latency_ms: float

@dataclass
class Packet:
    packet_id: str
    source_user_id: str
    destination_user_id: str
    station_source: str
    station_dest: str
    type: str
    time_sent_from_source_ms: float
    payload_data_base64: str
    payload_size_byte: int
    service_qos: QoS
    current_holding_node_id: str
    next_hop_node_id: str
    priority_level: int
    max_acceptable_latency_ms: float
    max_acceptable_loss_rate: float
    analysis_data: AnalysisData
    use_rl: bool
    ttl: int
    
    # Optional fields với giá trị mặc định
    acknowledged_packet_id: Optional[str] = None
    path_history: List[str] = field(default_factory=list)
    hop_records: List[HopRecord] = field(default_factory=list)
    accumulated_delay_ms: float = 0.0
    dropped: bool = False
    drop_reason: Optional[str] = None

    def __post_init__(self):
        if self.path_history is None:
            self.path_history = []
        if self.hop_records is None:
            self.hop_records = []

    def add_hop_record(self, hop_record: HopRecord):
        self.hop_records.append(hop_record)
        self.path_history.append(hop_record.to_node_id)
        self.accumulated_delay_ms += hop_record.latency_ms